File: commons-aws/src/test/java/test/pl/chilldev/commons/aws/MessageHandlerTest.java
/*
 * This file is part of the ChillDev-Commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2017 - 2018 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package test.pl.chilldev.commons.aws;

import java.util.function.Consumer;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import pl.chilldev.commons.aws.MessageHandler;

@ExtendWith(MockitoExtension.class)
public class MessageHandlerTest
{
    @Mock
    private Consumer<Integer> messageHandler;

    @Test
    public void handle() throws JsonProcessingException
    {
        ObjectMapper objectMapper = new ObjectMapper();

        MessageHandler<Integer> messageHandler = new MessageHandler<>(
            objectMapper,
            this.messageHandler,
            Integer.TYPE
        );

        Integer message = 44;
        messageHandler.handle(objectMapper.writeValueAsString(message));

        Mockito.verify(this.messageHandler).accept(message);
    }

    @Test
    public void handleInvalidJson()
    {
        ObjectMapper objectMapper = new ObjectMapper();

        MessageHandler<Integer> messageHandler = new MessageHandler<>(
            objectMapper,
            this.messageHandler,
            Integer.TYPE
        );

        Assertions.assertThrows(
            IllegalArgumentException.class,
            () -> messageHandler.handle("test"),
            "MessageHandler.handle() should throw exception when passed data is of different type."
        );
    }
}

File: commons-aws/src/test/kotlin/test/pl/wrzasq/commons/aws/messaging/MessageHandlerTest.kt
/*
 * This file is part of the pl.wrzasq.commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2017 - 2019, 2021 - 2022 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package test.pl.wrzasq.commons.aws.messaging

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.runs
import io.mockk.verify
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import pl.wrzasq.commons.aws.messaging.MessageHandler

@ExtendWith(MockKExtension::class)
class MessageHandlerTest {
    @MockK
    lateinit var messageHandler: (Int) -> Unit

    private val json = Json.Default

    @Test
    fun handle() {
        every { messageHandler(any()) } just runs

        val messageHandler = MessageHandler(
            json,
            messageHandler,
            json.serializersModule.serializer()
        )
        val message = 44
        messageHandler.handle(json.encodeToString(message))

        verify { messageHandler(message) }
    }

    @Test
    fun handleInvalidJson() {
        val messageHandler = MessageHandler(
            json,
            messageHandler,
            json.serializersModule.serializer()
        )
        assertThrows<IllegalArgumentException> { messageHandler.handle("test") }
    }
}
